Runtime optional casts.

atrick opened the pull request:
@jckarter Let me know if you foresee any problem with this change in cast behavior.

atrick committed:
Add an Optional metadata kind for runtime casts.

Reuses the enum metadata layout and builder because most of the logic is
also required for Optional (generic arg and payload). We may want to
optimize this at some point (Optional doesn't have a Parent), but I
don't see much opportunity.

Note that with this approach there will be no change in metadata layout.
Changing the kind still breaks the ABI of course.

Also leaves the MirrorData summary string as "(Enum Value)". We should
consider changing it.

gribozavr commented on test/1_stdlib/Runtime.swift:
```diff
@@ -823,6 +823,48 @@ Runtime.test("casting AnyObject to class metatypes") {
+  let ssx : SomeClass?? = sx
+  print(anyToAny(ssx, Optional<SomeClass>.self)! === x)
+  print(anyToAny(x, Optional<Optional<SomeClass>>.self)!! === x)
+  print(anyToAnyOrNil(ni, Int.self) == nil)
```
print?  Shouldn't it use expectEqual?

atrick replied:
Oops. I decided to move these all to an Interpreter test case, rewrote them all as prints, then changed my mind.

gribozavr commented on test/1_stdlib/Print.swift:
```diff
@@ -169,7 +169,7 @@ func test_StdlibTypesPrinted() {
     printedIs(optionalString, "nil")
     optionalString = "meow"
-    printedIs(optionalString, "Optional(\"meow\")")
+    printedIs(optionalString, "meow")
```
I find this change surprising.  Why is it an improvement?

jckarter replied:
Agreed, this looks fishy. @atrick can you look into why this happens?

atrick replied:
Sure, this is the behavior change that I wanted reviewed.

This happens because _print_unlocked first checks whether the value is convertible to CustomDebugStringConvertible. String is. With my change we correctly recognize that and just print the string itself. Previously we would fall into ad-hoc printing that reflects the type.

I could modify stdlib to special case printing of optionals, but I don't know why we would want to preserve the old behavior.

atrick replied:
@slavapestov may want to review this.

atrick committed:
Add Builtin.isOptional.

There was previously no way to detect a type that is nominally
Optional at runtime. The standard library, namely OutputStream, needs
to handle Optionals specially in order to cirumvent conversion to the
Optional's wrapped type. This should be done with conditional
conformance, but until that feature is available, Builtin.isOptional
will serve as a useful crutch.

atrick committed:
Fix dynamic runtime casts of Optionals.

Fixes <rdar://23122310> Runtime dynamic casts...

This makes runtime dynamic casts consistent with language rules, and
consequently makes specialization of generic code consistent with an
equivalent nongeneric implementation.

The runtime now supports casts from Optional<T> to U. Naturally the
cast fails on nil source, but otherwise succeeds if T is convertible to
U.

When casting T to Optional<U> the runtime succeeds whenever T is
convertible to U and simply wraps the result in an Optional.

To greatly simplify the runtime, I am assuming that
target-type-specific runtime cast entry points
(e.g. swift_dynamicCastClass) are never invoked with an optional
source. This assumption is valid for the following reasons. At the
language level optionals must be unwrapped before downcasting (via
as[?!]), so we only need to worry about SIL and IR lowering.  This
implementation assumes (with asserts) that:

- SIL promotion from an address cast to a value casts should only happen
  when the source is nonoptional. Handling optional unwrapping in SIL
  would be too complicated because we need to check for Optional's own
  conformances. (I added a test case to ensure this promotion does not
  happen). This is not an issue for unchecked_ref_cast, which
  implicitly unwraps optionals, so we can promote those!

- IRGen lowers unchecked_ref_cast (Builtin.castReference) directly to
  a bitcast (will be caught by asserts).

- IRGen continues to emit the generic dynamicCast entry point for
  address-casts (will be caught by asserts).
